src/audio/module_adapter/module/generic.c

```diff
@@ -411,7 +413,45 @@ int mod_free(struct processing_module *mod, const void *ptr)
 
 	return -EINVAL;
 }
-EXPORT_SYMBOL(mod_free);
+EXPORT_SYMBOL(z_impl_mod_free);
+
+#ifdef CONFIG_USERSPACE
+const void *z_vrfy_mod_fast_get(struct processing_module *mod, const void * const dram_ptr,
+				size_t size)
+{
+	struct module_resources *res = &mod->priv.resources;
+
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(mod, sizeof(*mod)));
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(res->heap, sizeof(*res->heap)));
+	K_OOPS(K_SYSCALL_MEMORY_READ(dram_ptr, size));
+
+	return z_impl_mod_fast_get(mod, dram_ptr, size);
+}
+#include <zephyr/syscalls/mod_fast_get_mrsh.c>
+
+void *z_vrfy_mod_alloc_ext(struct processing_module *mod, uint32_t flags, size_t size,
+			   size_t alignment)
+{
+	struct module_resources *res = &mod->priv.resources;
```

Comment:
We dereference mod before the access check. This can be used to have kernel make a memory read from arbitrary memory location from a unprivileged thread. The value cannot be retrieved, but a hw read can have side-effects, so better do the checks first.